src/profile-logic/profile-data.js

```diff
@@ -2911,3 +2915,419 @@ export function hasThreadKeys(
+  thread: Thread,
+  mapForSampleSelfStacks: Uint32Array,
+  mapForBacktraceSelfStacks: Uint32Array
+): Thread {
+  const {
+    samples: oldSamples,
+    markers: oldMarkers,
+    jsAllocations: oldJsAllocations,
+    nativeAllocations: oldNativeAllocations,
+  } = thread;
+
+  // Samples
+  const samples = {
+    ...oldSamples,
+    stack: oldSamples.stack.map((oldStackIndex) =>
+      oldStackIndex === null ? null : mapForSampleSelfStacks[oldStackIndex]
+    ),
+  };
+
+  // Markers
+  function replaceStackReferenceInPayload(
+    oldData: MarkerPayload
+  ): MarkerPayload {
+    if (oldData && 'cause' in oldData && oldData.cause) {
+      // Replace the cause with the right stack index.
+      // Use (...: any) because our current version of Flow has trouble with
+      // the object spread operator.
+      const newStack = mapForBacktraceSelfStacks[oldData.cause.stack];
+      return ({
+        ...oldData,
+        cause: {
+          ...oldData.cause,
+          stack: newStack,
+        },
+      }: any);
+    }
+    return oldData;
+  }
+  const markers = {
+    ...oldMarkers,
+    data: oldMarkers.data.map(replaceStackReferenceInPayload),
+  };
+
+  // JS allocations
+  let jsAllocations;
+  if (oldJsAllocations !== undefined) {
+    jsAllocations = {
+      ...oldJsAllocations,
+      stack: oldJsAllocations.stack.map((oldStackIndex) =>
+        oldStackIndex === null ? null : mapForBacktraceSelfStacks[oldStackIndex]
+      ),
+    };
+  }
+
+  // Native allocations
+  let nativeAllocations;
+  if (oldNativeAllocations !== undefined) {
+    nativeAllocations = {
+      ...oldNativeAllocations,
+      stack: oldNativeAllocations.stack.map((oldStackIndex) =>
+        oldStackIndex === null ? null : mapForBacktraceSelfStacks[oldStackIndex]
+      ),
+    };
+  }
+
+  return { ...thread, samples, markers, jsAllocations, nativeAllocations };
+}
+
+/**
+ * Creates a new thread with modified frame and stack tables for "nudged" return addresses:
+ * All return addresses are moved backwards by one byte, to point into the "call"
+ * instruction. This allows symbolication to obtain accurate line numbers and inline frames
+ * for these frames.
+ * Addresses which were sampled from the instruction pointer register remain unchanged.
+ * Non-native frames (i.e profiler label frames or JS frames) also remain unchanged.
+ *
+ * This function is called at the end of profile processing. In the Gecko profile format,
+ * caller frame addresses are return addresses. In the processed profile format, caller
+ * frame addresses point at/into the call instruction.
+ *
+ * # Motivation
+ *
+ * When the profiler captures a stack, the frame addresses in the stack come from two
+ * different sources:
+ *
+ *  1. The top frame comes from the instruction pointer register. The register contains the
+ *     address of the instruction that is currently executing.
+ *  2. The other frames come from stack walking. During stack walking, we record the
```

Comment:
I know of only one counter-example: the breakpad minidump stackwalkers.

I've changed it to "Stack walkers calculate the return addresses" because even if a stack walker records the nudged address, it will still have calculated the return address first.

Reply:
I've also added a comment that the Gecko profile format assumes return addresses.